docs/Implementing-easyblocks.rst

```diff
@@ -0,0 +1,486 @@
+            """Constructor for Example easyblock, initialises class variables."""
+
+            # call out to original constructor first, so 'self' (i.e. the class instance) is initialised
+            super(EB_Example, self).__init__(*args, **kwargs)
+
+            # initialise class variables
+            self.var1 = None
+            self.var2 = None
+
+
+.. _implementing_easyblocks_files:
+
+Reading/writing/copying/patching files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+File manipulation is a common use case for implementing easyblocks, hence the EasyBuild framework provides a
+number of useful functions related to this, including:
+
+* ``read_file(<path>)``: read file at a specified location and returns its contents
+
+* ``write_file(<path>, <text>)`` at a specified location with provided contents;
+  to append to an existing file, use ``append=True`` as an extra argument
+
+* ``copy_file(<src>, <dest>)`` to copy an existing file
+
+* ``apply_regex_substitutions(<path>, <list of regex substitions>)`` to patch an existing file
+
+All of these functions are provided by the ``easybuild.tools.filetools`` module.
+
+
+.. _implementing_easyblocks_commands:
+
+Executing (interactive) shell commands
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For executing shell commands two functions are provided by the ``easybuild.tools.run`` module:
+
+* ``run_cmd(<cmd>)`` to run a non-interactive shell command
+
+* ``run_cmd_qa(<cmd>, <dict with questions & answers>)`` to run an interactive shell command
+
+Both of these accept a number of optional arguments:
+
+* ``simple=True`` to just return ``True`` or ``False`` to indicate a successful execution,
+  rather than the default return value, i.e., a tuple that provides the command output and the exit code (in that order)
+
+* ``path=<path>`` to run the command in a specific subdirectory
+
+The ``run_cmd_qa`` function takes two additional specific arguments:
+
+* ``no_qa=<list>`` to specify a list of patterns to recognize non-questions
+
+* ``std_qa=<dict>`` to specify patterns for common questions and the matching answer
+
+
+.. _implementing_easyblocks_sanity_check:
+
+Custom (default) sanity check
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+For software-specific easyblocks, a custom sanity check is usually included to verify that the installation was
+successful or not.
+
+This is done by redefining the ``sanity_check_step`` method in the easyblock. For example:
+
+.. code:: python
+
+    from easybuild.framework.easyblock import EasyBlock
+
+    class EB_Example(EasyBlock):
+        """Custom easyblock for Example"""
+
+        def sanity_check_step(self):
+            """Custom sanity check for Example."""
+
+            custom_paths = {
+                'files': ['bin/example'],
+                'dirs': [],
+            }
+            custom_commands = ['example --version']
+            super(EB_Example, self).sanity_check_step(custom_paths=custom_paths, custom_commands=custom_commands)
```

Comment:
Make it more obvious what this line is? There might be people without OOP knowledge or Python knowledge or both, so including some comment on the code  or in the text might be useful